### DISCUSSION TITLE
understanding the score parameter

### DISCUSSION THREAD

**Discussion by danlangford** [pappde/bmai#6] [Q&A] [answered]
so currently BMAI had a tendency to surrender more games then it probably should and I think that I am not using `getaction` correctly. better stated i think that I am not helping bmai understand the dice that it has already taken. 

from the README

>Here is a sample that the web interface has created. It describes a situation where BMAI has 31 points and 2 dice (4-sided showing 1 and a 10-sided showing 1).  Its opponent has 40 points and 2 dice (20-sided speed showing 20, and 20-sided poison showing 2).  It is set to 2-ply search (use one or two, anything else is too expensive).  At 2-ply it uses simulations to evaluate the position.  It runs 150 simulations for each possible move, but will avoid running more than 1500 simulations total. The "getaction" command tells BMAI to run and output the desired action.
> ```
> game
> fight
> player 0 2 31
> 4:1
> 10:1
> player 1 2 40
> z20:20
> p20:2
> ply 2
> sims 150
> maxbranch 1500
> getaction
> ```

I am wondering about the `score` parameter. I can see that `BMC_Player` has a `float m_score;` and every time dice are captured that score is adjusted. I can see how each dice can produce a different score for when its is owned by its player vs when it has been captured. What i need to determine is in the line:
```
player 0 2 31
```
does the `31` mean that this player has captured 31 sides so far and if you calculate in this players remaining dice at half value
```
4:1
10:1
```
that is another 7 sides for a total score of 38 thus far?
Or is the 31 the total score for the player and adding the two additional dice will not alter the score so 31 would represent 24 sides from an opponent and 7 sides from this players remaining dice. 

again, as we discussed in #3 you dont have to feel obligated to dig in and figure this out @hamstercrack but if you happen to recall or want to you can reply. if i find the answer i will post it here

**Comment by danlangford** [comment, marked as answer]
so what i can tell from this code:
```cpp
// parse players
	for (i=0; i<BMD_MAX_PLAYERS; i++)
	{
		Read();
		INT p, dice;
		F32	score;
		if (sscanf(line, "player %d %d %f", &p, &dice, &score)<3 || p!=i)
			BMF_Error( "missing player: %s", line );
		ParsePlayer(p, dice);
		// don't clobber score in INITIATIVE phases
		if (g_game.GetPhase()!=BME_PHASE_INITIATIVE && g_game.GetPhase()!=BME_PHASE_INITIATIVE_CHANCE && g_game.GetPhase()!=BME_PHASE_INITIATIVE_FOCUS)
			g_game.GetPlayer(p)->m_score = score;
		// if this is still preround, then print all dice (i.e. including those not ready)
		if (g_game.IsPreround())
			g_game.GetPlayer(p)->DebugAllDice();
		else
			g_game.GetPlayer(p)->Debug();
	}
```

the `ParsePlayer()` is called there in the middle which in turn calls `BMC_Player::OnDiceParsed()`. That function sets the `m_score` to zero and then increases the `m_score` for each dice parsed meaning the dice that the player currently controls
```cpp
		m_score += m_die[i].GetScore(true);
```
so i learned: `m_score` includes dice under control. 

as a side note `m_score` also includes the score of dice we capture
```cpp
void BMC_Player::OnDieCaptured(BMC_Die *_die)
{
	m_score += _die->GetScore(false);
}
```

but back to `//parse players`

but after all that is done, since our phase is `BME_PHASE_FIGHT` i think we fall into
```cpp
		if (g_game.GetPhase()!=BME_PHASE_INITIATIVE && g_game.GetPhase()!=BME_PHASE_INITIATIVE_CHANCE && g_game.GetPhase()!=BME_PHASE_INITIATIVE_FOCUS)
			g_game.GetPlayer(p)->m_score = score;
```
where the `m_score` is overwritten completely. so i think i have concluded that the score i submit needs to be the entire score, both captured dice and remaining dice. 

thats a good thing :-) i have that number from the API.

**Reply by pappde** [reply]
Score:
The score parameter in the input file is exactly what the UI (website) reported as the current score. It should be the correct Button Men scoring: if the player has captured 30 sides and has 14 sides still in play, then the score should be reported as 37.  Note that m_score gets set in two different ways depending on state:
- initiative: BMC_Player::OnDiceParsed() - computes score using 0.5*own dice sides (since UI did not report this)
- playing state: m_score taken from input file (since input file does not detail captured dice so BMAI cannot compute)
